File: backend/models/ProjectRequest.js
const mongoose = require('mongoose');

const projectRequestSchema = new mongoose.Schema({
  name: { type: String, required: true, trim: true },
  clientName: { type: String, required: true, trim: true },
  totalCost: { type: Number, default: 0 },
  initialPayment: { type: Number, default: 0 },
  milestones: [{ type: String, trim: true }],
  startDate: { type: Date, required: true },
  expectedCompletionDate: { type: Date, required: true },
  projectStatus: {
    type: String,
    enum: ['planned', 'inProgress', 'completed'],
    default: 'planned',
  },
  requestedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  requestedByName: { type: String, default: '' },
  approvals: {
    developer: { type: Boolean, default: null },
    boss: { type: Boolean, default: null },
    accountant: { type: Boolean, default: null },
    developerUid: { type: String, default: null },
    bossUid: { type: String, default: null },
    accountantUid: { type: String, default: null },
    developerAt: { type: Date, default: null },
    bossAt: { type: Date, default: null },
    accountantAt: { type: Date, default: null },
  },
  status: {
    type: String,
    enum: ['pending', 'partiallyApproved', 'approved', 'rejected'],
    default: 'pending',
  },
  proofPath: { type: String, default: null },
  bankAccountId: { type: mongoose.Schema.Types.ObjectId, ref: 'BankAccount', default: null },
  projectId: { type: mongoose.Schema.Types.ObjectId, ref: 'Project', default: null },
  createdAt: { type: Date, default: Date.now },
});

projectRequestSchema.methods.toPublicJSON = function () {
  return {
    id: this._id.toString(),
    name: this.name,
    clientName: this.clientName,
    totalCost: this.totalCost,
    initialPayment: this.initialPayment,
    milestones: this.milestones || [],
    startDate: this.startDate.toISOString(),
    expectedCompletionDate: this.expectedCompletionDate.toISOString(),
    projectStatus: this.projectStatus,
    requestedBy: this.requestedBy.toString(),
    requestedByName: this.requestedByName,
    approvals: {
      developer: this.approvals.developer,
      boss: this.approvals.boss,
      accountant: this.approvals.accountant,
      developerUid: this.approvals.developerUid,
      bossUid: this.approvals.bossUid,
      accountantUid: this.approvals.accountantUid,
      developerAt: this.approvals.developerAt ? this.approvals.developerAt.toISOString() : null,
      bossAt: this.approvals.bossAt ? this.approvals.bossAt.toISOString() : null,
      accountantAt: this.approvals.accountantAt ? this.approvals.accountantAt.toISOString() : null,
    },
    status: this.status,
    proofPath: this.proofPath || null,
    bankAccountId: this.bankAccountId ? this.bankAccountId.toString() : null,
    projectId: this.projectId ? this.projectId.toString() : null,
    createdAt: this.createdAt.toISOString(),
  };
};

module.exports = mongoose.model('ProjectRequest', projectRequestSchema);

File: lib/models/project_request.dart
import 'payment.dart';
import 'project.dart';

class ProjectRequest {
  final String id;
  final String name;
  final String clientName;
  final double totalCost;
  final double initialPayment;
  final List<String> milestones;
  final DateTime startDate;
  final DateTime expectedCompletionDate;
  final ProjectStatus projectStatus;
  final String requestedBy;
  final String requestedByName;
  final PaymentApproval approvals;
  final PaymentStatus status;
  final String? projectId;
  final String? proofPath;
  final DateTime createdAt;

  ProjectRequest({
    required this.id,
    required this.name,
    required this.clientName,
    required this.totalCost,
    required this.initialPayment,
    required this.milestones,
    required this.startDate,
    required this.expectedCompletionDate,
    required this.projectStatus,
    required this.requestedBy,
    this.requestedByName = '',
    required this.approvals,
    required this.status,
    this.projectId,
    this.proofPath,
    required this.createdAt,
  });

  factory ProjectRequest.fromMap(Map<dynamic, dynamic> map) {
    final approvals = PaymentApproval.fromMap(
      (map['approvals'] as Map<dynamic, dynamic>?) ?? {},
    );
    return ProjectRequest(
      id: map['id'] ?? '',
      name: map['name'] ?? '',
      clientName: map['clientName'] ?? '',
      totalCost: (map['totalCost'] as num?)?.toDouble() ?? 0.0,
      initialPayment: (map['initialPayment'] as num?)?.toDouble() ?? 0.0,
      milestones: (map['milestones'] as List<dynamic>?)
              ?.map((m) => m.toString())
              .toList() ??
          [],
      startDate: DateTime.tryParse(map['startDate']?.toString() ?? '') ??
          DateTime.now(),
      expectedCompletionDate:
          DateTime.tryParse(map['expectedCompletionDate']?.toString() ?? '') ??
              DateTime.now(),
      projectStatus: _parseProjectStatus(map['projectStatus']?.toString()),
      requestedBy: map['requestedBy'] ?? '',
      requestedByName: map['requestedByName'] ?? '',
      approvals: approvals,
      status: _computeStatus(approvals),
      projectId: map['projectId'] as String?,
      proofPath: map['proofPath'] as String?,
      createdAt: DateTime.tryParse(map['createdAt']?.toString() ?? '') ??
          DateTime.now(),
    );
  }

  static ProjectStatus _parseProjectStatus(String? s) {
    switch (s) {
      case 'inProgress':
        return ProjectStatus.inProgress;
      case 'completed':
        return ProjectStatus.completed;
      default:
        return ProjectStatus.planned;
    }
  }

  static PaymentStatus _computeStatus(PaymentApproval approvals) {
    if (approvals.isRejected) return PaymentStatus.rejected;
    if (approvals.isFullyApproved) return PaymentStatus.approved;
    if (approvals.hasAnyApproval) return PaymentStatus.partiallyApproved;
    return PaymentStatus.pending;
  }

  String get statusDisplayName {
    switch (status) {
      case PaymentStatus.pending:
        return 'Pending Approval';
      case PaymentStatus.partiallyApproved:
        return 'Partially Approved';
      case PaymentStatus.approved:
        return 'Approved';
      case PaymentStatus.rejected:
        return 'Rejected';
    }
  }
}
